Games/2048/2048.cpp

```diff
@@ -0,0 +1,351 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "2048.h"
+#include <LibCore/ConfigFile.h>
+#include <LibGUI/FontDatabase.h>
+#include <LibGUI/MessageBox.h>
+#include <LibGUI/Painter.h>
+#include <LibGfx/Bitmap.h>
+#include <LibGfx/Font.h>
+#include <LibGfx/Palette.h>
+#include <stdlib.h>
+#include <time.h>
+
+TwentyFortyEightGame::TwentyFortyEightGame()
+{
+    set_font(GUI::FontDatabase::the().get_by_name("Liza Regular"));
+    srand(time(nullptr));
+    reset();
+}
+
+TwentyFortyEightGame::~TwentyFortyEightGame()
+{
+}
+
+template<typename Board>
+void TwentyFortyEightGame::add_tile(Board& board, int max_tile_value)
+{
+    int row;
+    int column;
+    do {
+        row = rand() % m_rows;
+        column = rand() % m_columns;
+    } while (board[row][column] != 0);
+
+    int value = rand() % max_tile_value;
+    value = round_up_to_power_of_two(value, max_tile_value);
+    board[row][column] = max(2, value);
+}
+
+void TwentyFortyEightGame::reset()
+{
+    auto initial_state = [&]() -> State {
+        State state;
+        state.board.resize(m_columns);
+        auto& board = state.board;
+        for (auto& row : board) {
+            row.resize(m_rows);
+            for (auto& j : row)
+                j = 0;
+        }
+
+        add_tile(state.board, m_starting_tile);
+        add_tile(state.board, m_starting_tile);
+
+        return state;
+    };
+
+    m_states.clear();
+    m_states.append(initial_state());
+
+    m_states.last().score_text = String::format("Score: %d", score());
+
+    update();
+}
+
+Gfx::IntRect TwentyFortyEightGame::score_rect() const
+{
+    int score_width = font().width(m_states.last().score_text);
+    return { 0, 2, score_width, font().glyph_height() };
+}
+
+static Vector<Vector<u32>> transpose(const Vector<Vector<u32>>& board)
+{
+    Vector<Vector<u32>> new_board;
+    auto result_row_count = board[0].size();
+    auto result_column_count = board.size();
+
+    new_board.resize(result_row_count);
+
+    for (size_t i = 0; i < board.size(); ++i) {
+        auto& row = new_board[i];
+        row.clear_with_capacity();
+        row.ensure_capacity(result_column_count);
+        for (auto& entry : board) {
+            row.append(entry[i]);
+        }
+    }
+
+    return move(new_board);
```

Comment:
I think the compiler will automatically move return values and you don't need to explicitly `move()` them.